appengine/pond_storage/rank.py

```diff
@@ -93,10 +99,36 @@ def create_ranking_match():
   # 4. Store match request in datastore.
   match = MatchRequest(entry_keys=entry_keys)
   match_key = match.put()
-  match_key_urlsafe = match_key.urlsafe()
-  logging.info('Created match request. matchKey:%s', match_key_urlsafe)
+  logging.info('Created match request. matchKey: %s', match_key)
   # 5. Return match request information.
-  return {'duck_list': duck_list, 'match_key': match_key_urlsafe}
+  return {'duck_list': duck_list, 'match_key': match_key.urlsafe()}
+
+def move_down_dummy_entry(dummy_entry, retries=5):
+  """Tries to move dummy entry down and/or deletes it if it is at bottom."""
+  @ndb.transactional(xg=True)
+  def swap_with_lower(lower_entry_key):
+    lower_entry = lower_entry_key.get()
+    # First confirm that lower entry rank is still farther down in leaderboard.
+    if not lower_entry and lower_entry.ranking > dummy_entry.key.get().ranking:
```

Comment:
One tweak then lgtm. I don't think we need the not in front of the lower_entry. So would be if lower_entry and lower_entry.ranking....

Reply:
The not is needed. I'm checking that the lower_entry exists inside the transaction.
I'm returning false if the entry does not exist (hence the not) or it is no longer lower.

Reply:
My concern here is that lower_entry gets changed to return something like None and then we try to call .ranking on a None type and get an error.

Reply:
I think you are referring to the outdated code. The current code is:
`if not lower_entry or lower_entry.ranking < dummy_entry.key.get().ranking:`
So ranking would not be called on a `None` type because `not lower_entry` would be `True`

The importance is that I want the if condition to **also** pass if `lower_entry` is `None` because I don't want to try to swap it if it doesn't exist anymore (vs if I had `lower_entry and` which would only prevent ranking being called on `None` type and **not** if `lower_entry` was `None`)

Reply:
ahhh you are right I was looking at outdated. And thanks for the explanation that makes sense : )